src/main/scala/mesosphere/marathon/core/election/CuratorElectionStream.scala

```diff
@@ -0,0 +1,231 @@
+          case LeadershipState.ElectedAsLeader =>
+            logger.info("Leader won.")
+          case LeadershipState.Standby(currentLeader) =>
+            logger.info(s"Leader defeated. Current leader: ${currentLeader.getOrElse("-")}")
+        }
+        e
+      }
+  }
+
+  @SuppressWarnings(Array("CatchThrowable"))
+  private class CuratorEventEmitter(
+      singleThreadExecutor: ExecutorService,
+      clientCloseable: LifeCycledCloseableLike[CuratorFramework],
+      zooKeeperLeaderPath: String,
+      hostPort: String,
+      sq: SourceQueueWithComplete[LeadershipState]) extends Cancellable {
+
+    val client = clientCloseable.closeable
+    private lazy val singleThreadEC: ExecutionContext = ExecutionContext.fromExecutor(singleThreadExecutor)
+    private lazy val leaderHostPortMetric: Timer = Metrics.timer(ServiceMetric, getClass, "current-leader-host-port")
+    private val curatorLeaderLatchPath = zooKeeperLeaderPath + "-curator"
+    private lazy val latch = new LeaderLatch(client, curatorLeaderLatchPath, hostPort)
+    @volatile private var isStarted = false
+    @volatile private var _isCancelled = false
+
+    /* It is important that we re-register our watch _BEFORE_ we get the current the current leader. In Zookeeper,
+     * watches are one-time use only.
+     *
+     * The timeline of events looks like this
+     *
+     * 1) We register a watch
+     * 2) We query the current leader
+     * 3) We receive an event (indicating child removal / addition)
+     * 4) Repeat step 1
+     *
+     * By re-registering the watch before querying the state, we will not miss out on the latest leader change.
+     */
+    def longPollLeaderChange(retries: Int = 0): Unit = Future {
+      try {
+        if (latch.getState == LeaderLatch.State.STARTED)
+          client.getChildren
+            .usingWatcher(new CuratorWatcher {
+              override def process(event: WatchedEvent): Unit =
+                if (!_isCancelled) longPollLeaderChange()
+            })
+            .forPath(curatorLeaderLatchPath)
+        emitLeader()
+      } catch {
+        case ex: KeeperException.NoNodeException if retries < 100 =>
+          // Wait for node to be created
+          logger.info("retrying")
+          Thread.sleep(retries * 10L)
+          longPollLeaderChange(retries + 1)
+        case ex: Throwable =>
+          sq.fail(ex)
+      }
+    }(singleThreadEC)
+
+    /**
+      * Emit current leader. Does not fail on connection error
+      */
+    private def emitLeader(): Unit = {
+      val currentLeader = leaderHostPortMetric.blocking {
+        if (client.getState == CuratorFrameworkState.STOPPED) {
+          None
+        } else {
+          try {
+            Some(latch.getLeader.getId)
+          } catch {
+            case ex: Throwable =>
+              logger.error("Error while getting current leader", ex)
+              None
+          }
+        }
+      }
+
+      currentLeader match {
+        case Some(`hostPort`) =>
```

Comment:
I don't quite get this. What is the type of `currentLeader`? What is `` `hostPort` ``? Just a name?

Reply:
`currentLeader` is the value we queried from ZooKeeper. `hostPort` is passed to the constructor above.

I'll remove the back ticks and replace it with an if statement to be more clear. Kind of a fringe Scala feature.

Reply:
(this code is equivalent)

```
      currentLeader match {
        case Some(leader) if leader == hostPort =>
          sq.offer(LeadershipState.ElectedAsLeader)
        case otherwise =>
          sq.offer(LeadershipState.Standby(otherwise))
      }
```

Reply:
Jeez. That is indeed some fringe feature 😉. Thanks for making the code more expressive.